File: py-cmd-cli/sections/virtual_device.py
# Yealink Provision CLI - Virtual Device Editor
from .api import api_url

from .configEditor import ConfigCLI
from .model import get_model

import cmd2
import requests

class VirtualDevice:
    def __init__(self, site_id, id, name, model_id, remark, create_date):
        self.id = id
        self.name = name
        self.site_id = site_id
        self.model_id = model_id
        self.remark = remark
        self.create_date = create_date
        
    def rename(self, name):
        # Rename the virtual device.
        r = requests.patch(api_url + '/sites/' + self.site_id + '/virtual_devices/' + self.id, json={
            'name': name
        })
        
        if r.status_code == 200:
            self.name = name
            return True
        
        return False
    
    def delete(self):
        # Delete the virtual device
        r = requests.delete(api_url + '/sites/' + self.site_id + '/virtual_devices/' + self.id)
        
        if r.status_code == 200:
            return True
        
        return False
    
def create_virtual_device(site_id, name, model_id, remark):
    # Create the new vdev
    r = requests.post(api_url + '/sites/' + site_id + '/virtual_devices', json={
        'name': name,
        'model_id': model_id,
        'description': remark,
    })
    
    if r.status_code == 200:
        return VirtualDevice(site_id, r.json()['id'], name, model_id, remark, r.json()['create_date'])
    else:
        print(r.text)
        
    return None

def get_virtual_devices(site_id):
    # Get all vdevs within the site
    r = requests.get(api_url + '/sites/' + site_id + '/virtual_devices')
    
    if r.status_code == 200:
        vdevs = []
        for vdev in r.json():
            remark = vdev['description'] if 'description' in vdev else "N/A"
            vdevs.append(VirtualDevice(site_id, vdev['id'], vdev['name'], vdev['model_id'], remark, vdev['create_date']))
        return vdevs
    
    return None

def get_virtual_device(site_id, id):
    # Get a specific device within a site.
    r = requests.get(api_url + '/sites/' + site_id + '/virtual_devices/' + id)
    
    if r.status_code == 200:
        vdev = r.json()
        remark = vdev['description'] if 'description' in vdev else "N/A"
        return VirtualDevice(site_id, vdev['id'], vdev['name'], vdev['model_id'], remark, vdev['create_date'])
    else:
        print(f"vdev: failed to get vdev {r.json()['message']}")
    
    return None

class VirtualDeviceEditCLI(cmd2.Cmd):
    """Yealink Provision CLI - Virtual Device Editor"""
    prompt = 'site-vdev...> '
    
    def __init__(self, site, virtual_device):
        self.site = site
        self.vdev = virtual_device
        super().__init__()
        
        self.prompt = f"site({self.site.name})-vdev({self.vdev.name})> "
        
    def do_rename(self, args):
        """Rename the vdev"""
        if len(args) == 0:
            print("Error: no new name specified.")
            return
        
        if self.vdev.rename(args):
            print("Device renamed.")
            self.prompt = f"site({self.site.name})-vdev({self.vdev.name})> "
        else:
            print("error: failed to rename device.")
        
    def do_show(self, args):
        """Show the virtual device information"""
        print(f"ID: {self.vdev.id}")
        print(f"Name: {self.vdev.name}")
        print(f"Model ID: {self.vdev.model_id}")
        print(f"Remark: {self.vdev.remark}")
        print(f"Create Date: {self.vdev.create_date}")
        
    def do_delete(self, args):
        """Delete this device."""
        if self.vdev.delete():
            print("Device deleted.")
            return True
        else:
            print("Error: device delete failed.")
            
    def do_config(self, args):
        """Enter the virtual device configuration"""
        ConfigCLI("device", self.vdev.id, self.vdev.name).cmdloop()
    
    
class VirtualDeviceCLI(cmd2.Cmd):
    """Yealink Provision - Virtual Device Manager"""
    prompt = 'site-vdev...> '
    def __init__(self, site):
        self.site = site
        super().__init__()
        
        self.prompt = f"site({self.site.name})-vdev> "
    
    def do_show(self, args):
        """List all virtual devices."""
        vdevs = get_virtual_devices(self.site.id)
        if vdevs is None:
            print("Error: failed to get vdevs.")
            return
        
        print("ID\t\tName\t\tModel\t\tRemark\t\tCreate Date")
        for vdev in vdevs:
            print(vdev.id + "\t" + vdev.name + "\t" + get_model(vdev.model_id).name + "\t\t" + vdev.remark + "\t" + vdev.create_date)

    def do_edit(self, args):
        """Edit a device"""
        if len(args) == 0:
            print("Error: no virtual device specified.")
            return
        
        vdev = get_virtual_device(self.site.id, args)
        if vdev is None:
            print("Error: failed to get virtual device.")
            return
        
        VirtualDeviceEditCLI(self.site, vdev).cmdloop()
        
    def do_create(self, args):
        """Create a new vdev"""
        name = input("Name: ")
        model_id = input("Model ID: ")
        remark = input("Remark: ")
        
        vdev = create_virtual_device(self.site.id, name, model_id, remark)
        if vdev is None:
            print("Error: failed to create new virtual device.")
            return
        
        print("Device created, entering editor.")
        VirtualDeviceEditCLI(self.site, vdev).cmdloop()

File: api-server/routes/virtual_device.js
// yealink-provision - Virtual Device Management API
// Cameron Fleming 2023

// These endpoints are used to manage virtual devices within sites.

import { customAlphabet } from "nanoid";
import { Router } from "express";

import {VirtualDevice, virtualDeviceSchema} from '../mongo/schemas/virtual_device.js';
import { Site } from '../mongo/schemas/site.js';
import { Model } from '../mongo/schemas/model.js';

import { logger } from '../index.js';

// Setup nanoid with custom alphabet
const nanoid = customAlphabet('1234567890abcdef', 8);

const router = Router({ mergeParams: true });

// This router expects to behind a site ID (i.e., /sites/:id/virtual_devices/:id)
// So any requests to the root (/sites/:id/virtual_devices) will return devices specific to the site.

// Get all virtual devices
router.get('/', async(req, res) => {
  let virtual_devices;
  virtual_devices = await VirtualDevice.find( {site_id: req.params.site} );
  res.json(virtual_devices);
})

// Get a specific virtual device
router.get('/:id', async(req, res) => {
  const virtual_device = await VirtualDevice.findOne({id: req.params.id, site_id: req.params.site});
  if (!virtual_device) {
    res.status(404).json({
      error: 'not_found',
      message: 'This virtual device does not exist.'
    })
    return;
  }
  
  res.json(virtual_device);
})

// Create a new virtual device
router.post('/', async(req, res) => {
  // Verify the required params are present, except the ID which is generated dynamically.
  if (!req.body.name || !req.body.model_id) {
    res.status(400).json({
      error: 'missing_params',
      message: 'This endpoint requires a name and a model_id.'
    })
    logger.warn("vdev: rejecting creation of new vdev, missing params.")
    return;
  }
  
  // Generate new ID
  req.body.id = nanoid(8);
  
  // Verify the site exists
  const site = Site.findOne({id: req.params.site});
  if (!site) {
    res.status(400).json({
      error: 'invalid_site',
      message: 'The provided site does not exist.'
    })
    logger.warn("vdev: rejecting creation of new vdev, site doesn't exist.")
    return;
  }
  
  // Verify the model exists
  const model = await Model.findOne({id: req.body.model_id});
  if (!model) {
    res.status(400).json({
      error: 'invalid_model',
      message: 'No model exists with that ID.'
    })
    logger.warn("vdev: rejecting creation of new vdev, model doesn't exist.")
    return;
  }
  
  // Create the new device.
  req.body.site_id = req.params.site;
  const virtual_device = new VirtualDevice(req.body);
  await virtual_device.save();
  
  if (!virtual_device) {
    res.status(500).json({
      error: 'failed_to_create_virtual_device',
      message: 'Failed to create the virtual device.'
    })
    logger.error("vdev: Creation of the vdev failed.")
    return;
  }
  logger.info(`vdev: new virtual device created: ${virtual_device.id} - ${virtual_device.name} on site ${site.name}!`)

  res.json(virtual_device);
})

// Rename a vdevice or change its site.
router.patch('/:id', async(req, res) => {
  const virtual_device = await VirtualDevice.findOne({ id: req.params.id, site_id: req.params.site });
  if (!virtual_device) {
    res.status(400).json({
      error: 'invalid_virtual_device',
      message: 'This virtual device does not exist, or is not a member of this site.'
    })
    return;
  }
  
  // Rename the virtual device
  if (req.body.name) {
    logger.info(`vdev: virtual device ${virtual_device.id} renamed from ${virtual_device.name} to ${req.body.name}`)
    virtual_device.name = req.body.name;
  }
  
  // Save the changes
  await virtual_device.save();
  res.json(virtual_device);
  
  logger.info("vdev: device updates saved.")
})

// Delete a virtual device
router.delete('/:id', async(req, res) => {
  const virtual_device = VirtualDevice.findOne({ id: req.params.id, site_id: req.params.site });
  if (!virtual_device) {
    res.status(400).json({
      status: 'invalid_virtual_device',
      message: 'This virtual device does not exist, or does not belong to this site.'
    })
    return;
  }
  
  // Delete the virtual device.
  await VirtualDevice.deleteOne({ id: req.params.id });
  
  res.json({
    status: 'virtual_device_deleted',
    message: 'The virtual device was deleted successfully.'
  })
  
  logger.info(`vdev: virtual device delete: ${virtual_device.id}`)
})

export default router;